File: sdk/swift/Sources/OpenAgentTrustRegistry/Registry.swift
import Foundation

public class OpenAgentTrustRegistryClient {
    private let mirrorUrl: URL
    private var manifest: RegistryManifest?
    private var revocations: RevocationList?
    private var lastFetchTime: Date?
    
    // Refresh cache every 15 minutes by default
    private let cacheTTL: TimeInterval = 15 * 60

    private init(mirrorUrl: URL) {
        self.mirrorUrl = mirrorUrl
    }

    /// Initialize a new Registry Client and aggressively fetch the initial state.
    public static func load(mirrorUrl: URL) async throws -> OpenAgentTrustRegistryClient {
        let registry = OpenAgentTrustRegistryClient(mirrorUrl: mirrorUrl)
        try await registry.refresh()
        return registry
    }

    /// Manually trigger a refresh of the cached registry state from the network.
    public func refresh() async throws {
        // Build the precise endpoint URLs
        let manifestUrl = mirrorUrl.appendingPathComponent("v1/registry")
        let revocationsUrl = mirrorUrl.appendingPathComponent("v1/revocations")
        
        async let manifestTask = URLSession.shared.data(from: manifestUrl)
        async let revocationsTask = URLSession.shared.data(from: revocationsUrl)
        
        let (manifestData, manifestResponse) = try await manifestTask
        let (revocationsData, revocationsResponse) = try await revocationsTask
        
        guard let httpManifest = manifestResponse as? HTTPURLResponse, httpManifest.statusCode == 200,
              let httpRevocations = revocationsResponse as? HTTPURLResponse, httpRevocations.statusCode == 200 else {
            throw OpenAgentTrustRegistryError.fetchFailed("Failed to fetch remote registry state (non-200 HTTP response).")
        }
        
        let decoder = JSONDecoder()
        manifest = try decoder.decode(RegistryManifest.self, from: manifestData)
        revocations = try decoder.decode(RevocationList.self, from: revocationsData)
        lastFetchTime = Date()
    }
    
    /// Verify an incoming agent attestation locally against the cached registry JSON.
    public func verify(attestationJws: String, expectedAudience: String, expectedNonce: String? = nil) async throws -> VerificationResult {
        
        // Auto-refresh the cache if it's considered stale
        if let lastFetch = lastFetchTime, Date().timeIntervalSince(lastFetch) > cacheTTL {
            try await refresh()
        }
        
        guard let manifest = manifest, let revocations = revocations else {
            throw OpenAgentTrustRegistryError.registryNotLoaded
        }
        
        return Verification.verifyAttestation(
            attestationJws: attestationJws,
            manifest: manifest,
            revocations: revocations,
            expectedAudience: expectedAudience,
            expectedNonce: expectedNonce
        )
    }
}

File: sdk/typescript/src/registry.ts
// sdk/typescript/src/registry.ts

import type { RegistryManifest, RevocationList, VerificationResult } from './types';
import { verifyAttestation } from './verify';
import { OpenAgentTrustRegistryError, verifyRegistryArtifacts } from './registry-artifacts';

export class OpenAgentTrustRegistry {
  private mirrorUrl: string;
  private manifest: RegistryManifest | null = null;
  private revocations: RevocationList | null = null;
  private lastFetchTime: number = 0;
  
  // Refresh cache every 15 minutes by default
  private readonly CACHE_TTL_MS = 15 * 60 * 1000; 

  private constructor(mirrorUrl: string) {
    this.mirrorUrl = mirrorUrl;
  }

  /**
   * Initialize a new Registry Client and fetch the initial state.
   */
  public static async load(mirrorUrl: string): Promise<OpenAgentTrustRegistry> {
    const registry = new OpenAgentTrustRegistry(mirrorUrl.replace(/\/$/, ''));
    await registry.refresh();
    return registry;
  }

  /**
   * Manually trigger a refresh of the cached registry state from the network.
   */
  public async refresh(): Promise<void> {
    try {
      const [manifestRes, revocationsRes] = await Promise.all([
        fetch(`${this.mirrorUrl}/v1/registry`),
        fetch(`${this.mirrorUrl}/v1/revocations`)
      ]);

      if (!manifestRes.ok || !revocationsRes.ok) {
        throw new OpenAgentTrustRegistryError(
          'fetch_failed',
          `Failed to fetch registry state: ${manifestRes.status} / ${revocationsRes.status}`
        );
      }

      const manifest = await manifestRes.json();
      const revocations = await revocationsRes.json();
      const verifiedState = verifyRegistryArtifacts(manifest, revocations);

      this.manifest = verifiedState.manifest;
      this.revocations = verifiedState.revocations;
      this.lastFetchTime = Date.now();

    } catch (err) {
      console.error('[OpenAgentTrustRegistry] Failed to refresh state', err);
      throw err;
    }
  }

  /**
   * Verify an incoming agent attestation JWS token locally against the loaded registry state.
   * 
   * @param attestationJws The raw string JWS token
   * @param expectedAudience The specific service origin (aud) expecting this token
   * @param expectedNonce Optional nonce if the service requires intra-service replay protection
   */
  public async verifyToken(
    attestationJws: string, 
    expectedAudience: string,
    expectedNonce?: string
  ): Promise<VerificationResult> {
    
    // Auto-refresh if cache is stale
    if (Date.now() - this.lastFetchTime > this.CACHE_TTL_MS) {
        await this.refresh();
    }

    if (!this.manifest || !this.revocations) {
        throw new OpenAgentTrustRegistryError('registry_not_loaded', 'Registry state not loaded');
    }

    const verifiedState = verifyRegistryArtifacts(this.manifest, this.revocations);
    this.manifest = verifiedState.manifest;
    this.revocations = verifiedState.revocations;

    return verifyAttestation(
        attestationJws, 
        this.manifest, 
        this.revocations, 
        expectedAudience, 
        expectedNonce
    );
  }
}
